cpp/src/arrow/filesystem/filesystem.h

```diff
@@ -519,6 +574,82 @@ Result<std::shared_ptr<FileSystem>> FileSystemFromUriOrPath(
 
 /// @}
 
+/// \defgroup filesystem-factory-registration Helpers for FileSystem registration
+///
+/// @{
+
+/// \brief Register a Filesystem factory
+///
+/// Support for custom Uri schemes can be added by registering a factory
+/// for the corresponding FileSystem.
+///
+/// \param[in] scheme a Uri scheme which the factory will handle.
+///            If a factory has already been registered for a scheme,
+///            the new factory will be ignored.
+/// \param[in] factory a function which can produce a FileSystem for Uris which match
+///            scheme.
+/// \param[in] finalizer a function which must be called to finalize the factory before
+///            the process exits, or nullptr if no finalization is necessary.
+ARROW_EXPORT void RegisterFileSystemFactory(std::string scheme,
+                                            FileSystem::Factory factory,
+                                            void finalizer() = NULLPTR);
+
+/// \brief Register Filesystem factories from a shared library
+///
+/// In addition to dynamically loading the indicated library, registries are merged if
+/// necessary (static linkage can produce isolated registries).
+ARROW_EXPORT Status LoadFileSystemFactories(const char* libpath);
```

Comment:
Can the docstring make the contract more explicit? i.e., IIUC, the library is supposed to instantiate static FileSystemRegistrar instances that automatically register the provided factories?

Reply:
Alright, I'll copy some of the description of what's expected in a library to this docstring